### ISSUE TITLE
concoct package not compatible with anvio 5.4 in anaconda?

### ISSUE THREAD

**Issue by najoshi** [merenlab/anvio#1144]
```
Anvi'o version ...............................: margaret (v5.4)
Profile DB version ...........................: 31
Contigs DB version ...........................: 12
Pan DB version ...............................: 13
Genome data storage version ..................: 6
Auxiliary data storage version ...............: 2
Structure DB version .........................: 1
```

Hello,

I'm a sys admin trying to install anvio 5.4 on an HPC system using anaconda3 (4.6.11). I used these commands from the installation page to install anvio:

conda create -y --name anvio5 python=3.6
conda install -y --name anvio5 -c bioconda -c conda-forge anvio=5.4.0

I had a user trying to run anvio and it gave them an error about not having the concoct package installed. I looked, and indeed it is not installed, which was odd. So then I tried installing it and conda wanted to remove the anvio package in order to install concoct. So it seems like the version of concoct (concoct-0.4.1-py36hd1ccf6d_0) is incompatible with anvio 5.4? Any idea on how to fix this issue?

- Nik.

**Comment by meren** [comment]
CONCOCT is a native component of anvi'o and is not an additional package.  Do you think the misunderstanding comes from that?

**Comment by ozcan** [comment]
Hi Nik,
Even though we are planning to change this pretty soon and use the concoct available on your system, at current stable release, anvi'o comes with concoct module. You do not have to install concoct. However, under some circumstances, this C extension cannot be loaded and used by Python. 

I think the only solution would be compiling this extension on your system/for your system and replacing the binary compiled by Anaconda.

Example steps to do that:
```
cd /tmp && mkdir anvio_build && cd anvio_build
wget https://files.pythonhosted.org/packages/d6/5a/093c8d13686532dab92c9fd431aaac768c982b99c77676224281a066df4b/anvio-5.4.tar.gz
python setup.py build_ext
```

After this step you will have the concoct module a path similar to below:
```
build/lib.macosx-10.7-x86_64-3.6/anvio/vbgmm.cpython-36m-darwin.so
```
you need to replace prebuilt `vbgmm*.so` file with the file above. This file located inside anvi'o, And the running the command below should either print the path of this file or print an error (error also contains the path)
```
python -c "import anvio.vbgmm; print(anvio.vbgmm.__file__)"
```

Best,

**Comment by splaisan** [comment]
I tried but it seems my numpy is playing jokes or it is something else...

```
(anvio5) u0002316@gbw-s-pacbio01:/tmp/anvio_build/anvio-5.5$ python setup.py build_ext
running build_ext
skipping './anvio/extensions/concoct/vbgmm.c' Cython extension (up-to-date)
building 'anvio.vbgmm' extension
gcc -pthread -B /opt/biotools/miniconda2/envs/anvio5/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include -I/opt/local/include/ -I/usr/local/include/ -I/opt/biotools/miniconda2/envs/anvio5/include/python3.6m -c ./anvio/extensions/concoct/vbgmm.c -o build/temp.linux-x86_64-3.6/./anvio/extensions/concoct/vbgmm.o
In file included from /opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include/numpy/ndarraytypes.h:1788:0,
                 from /opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include/numpy/ndarrayobject.h:18,
                 from /opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from ./anvio/extensions/concoct/vbgmm.c:593:
/opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it by " \
  ^
gcc -pthread -B /opt/biotools/miniconda2/envs/anvio5/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include -I/opt/local/include/ -I/usr/local/include/ -I/opt/biotools/miniconda2/envs/anvio5/include/python3.6m -c ./anvio/extensions/concoct/c_vbgmm_fit.c -o build/temp.linux-x86_64-3.6/./anvio/extensions/concoct/c_vbgmm_fit.o
./anvio/extensions/concoct/c_vbgmm_fit.c:14:28: fatal error: gsl/gsl_vector.h: No such file or directory
compilation terminated.
error: command 'gcc' failed with exit status 1
```

**Comment by ozcan** [comment]
Sorry, I forgot to mention in the previous comment the build dependencies are different from the runtime dependencies, You need the things below:
    - numpy
    - cython
    - gsl
It seems you already have numpy (the warning is not really important). You need to install the rest using `conda install cython gsl`

**Comment by splaisan** [comment]
not yet there :-) the gsl is not found!

I did the two installs as advised then expanded fresh fro the tar.gz and tried again

```
(anvio5) u0002316@gbw-s-pacbio01:/tmp/anvio_build/anvio-5.5$ python setup.py build_ext
running build_ext
skipping './anvio/extensions/concoct/vbgmm.c' Cython extension (up-to-date)
building 'anvio.vbgmm' extension
creating build
creating build/temp.linux-x86_64-3.6
creating build/temp.linux-x86_64-3.6/anvio
creating build/temp.linux-x86_64-3.6/anvio/extensions
creating build/temp.linux-x86_64-3.6/anvio/extensions/concoct
gcc -pthread -B /opt/biotools/miniconda2/envs/anvio5/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include -I/opt/local/include/ -I/usr/local/include/ -I/opt/biotools/miniconda2/envs/anvio5/include/python3.6m -c ./anvio/extensions/concoct/vbgmm.c -o build/temp.linux-x86_64-3.6/./anvio/extensions/concoct/vbgmm.o
In file included from /opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include/numpy/ndarraytypes.h:1822:0,
                 from /opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from ./anvio/extensions/concoct/vbgmm.c:593:
/opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it with " \
  ^
gcc -pthread -B /opt/biotools/miniconda2/envs/anvio5/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/biotools/miniconda2/envs/anvio5/lib/python3.6/site-packages/numpy/core/include -I/opt/local/include/ -I/usr/local/include/ -I/opt/biotools/miniconda2/envs/anvio5/include/python3.6m -c ./anvio/extensions/concoct/c_vbgmm_fit.c -o build/temp.linux-x86_64-3.6/./anvio/extensions/concoct/c_vbgmm_fit.o
./anvio/extensions/concoct/c_vbgmm_fit.c:14:28: fatal error: gsl/gsl_vector.h: No such file or directory
compilation terminated.
error: command 'gcc' failed with exit status 1
(anvio5) u0002316@gbw-s-pacbio01:/tmp/anvio_build/anvio-5.5$ conda list
# packages in environment at /opt/biotools/miniconda2/envs/anvio5:
#
# Name                    Version                   Build  Channel
anvio                     5.5.0                         0    bioconda
anvio-minimal             5.5.0            py36hf3f1cc3_0    bioconda
appdirs                   1.4.3                      py_1    conda-forge
asn1crypto                0.24.0                py36_1003    conda-forge
atk                       2.25.90           hb9dd440_1002    conda-forge
attrs                     19.1.0                     py_0    conda-forge
bcftools                  1.9                  ha228f0b_3    bioconda
biopython                 1.73             py36h14c3975_0    conda-forge
blas                      2.8                    openblas    conda-forge
blast                     2.5.0                hc0b0e79_3    bioconda
blast-legacy              2.2.26                        1    bioconda
boost                     1.69.0          py36h8619c78_1001    conda-forge
boost-cpp                 1.69.0            h11c811c_1000    conda-forge
bottle                    0.12.13                    py_1    conda-forge
bowtie2                   2.3.5            py36he860b03_0    bioconda
bwa                       0.7.17               h84994c4_5    bioconda
bzip2                     1.0.6             h14c3975_1002    conda-forge
ca-certificates           2019.3.9             hecc5488_0    conda-forge
cairo                     1.16.0            ha4e643d_1000    conda-forge
centrifuge                1.0.4_beta      py36pl526he941832_2    bioconda
certifi                   2019.3.9                 py36_0    conda-forge
cffi                      1.12.3           py36h8022711_0    conda-forge
chardet                   3.0.4                 py36_1003    conda-forge
cherrypy                  8.0.0                    py36_0    conda-forge
colored                   1.3.93                     py_0    conda-forge
configargparse            0.13.0                     py_1    conda-forge
cryptography              2.6.1            py36h72c5cf5_0    conda-forge
curl                      7.64.1               hf8cf82a_0    conda-forge
cycler                    0.10.0                     py_1    conda-forge
cython                    0.29.7           py36he1b5a44_0    conda-forge
datrie                    0.7.1            py36h14c3975_0    conda-forge
dbus                      1.13.6               he372182_0    conda-forge
diamond                   0.9.21                        1    bioconda
django                    2.0.8                    py36_0    conda-forge
docutils                  0.14                  py36_1001    conda-forge
ete3                      3.1.1                    py36_0    bioconda
expat                     2.2.5             hf484d3e_1002    conda-forge
fontconfig                2.13.1            he4413a7_1000    conda-forge
freetype                  2.10.0               he983fc9_0    conda-forge
gdk-pixbuf                2.32.2                        1    bioconda
gettext                   0.19.8.1          hc5be6a0_1002    conda-forge
glib                      2.58.3            hf63aee3_1001    conda-forge
gobject-introspection     1.58.2          py36h2da5eee_1000    conda-forge
graphite2                 1.3.13            hf484d3e_1000    conda-forge
gsl                       2.4               h294904e_1006    conda-forge
gst-plugins-base          1.14.4            hdf3bae2_1001    conda-forge
gstreamer                 1.14.4            h66beb1c_1001    conda-forge
gtk2                      2.24.31           hb68c50a_1001    conda-forge
h5py                      2.9.0           nompi_py36hf008753_1102    conda-forge
harfbuzz                  2.4.0                h37c48d4_0    conda-forge
hdf5                      1.10.4          nompi_h3c11f04_1106    conda-forge
hmmer                     3.2.1                hf484d3e_1    bioconda
htslib                    1.9                  ha228f0b_7    bioconda
icu                       58.2              hf484d3e_1000    conda-forge
idna                      2.7                   py36_1002    conda-forge
illumina-utils            2.6                        py_0    bioconda
intel-openmp              2019.3                      199  
iqtree                    1.6.10               he860b03_0    bioconda
jpeg                      9c                h14c3975_1001    conda-forge
jsonschema                3.0.1                    py36_0    conda-forge
kiwisolver                1.1.0            py36hc9558a2_0    conda-forge
krb5                      1.16.3            h05b26f9_1001    conda-forge
libblas                   3.8.0                8_openblas    conda-forge
libcblas                  3.8.0                8_openblas    conda-forge
libcurl                   7.64.1               hda55be3_0    conda-forge
libdeflate                1.0                  h14c3975_1    bioconda
libedit                   3.1.20170329      hf8c457e_1001    conda-forge
libffi                    3.2.1             he1b5a44_1006    conda-forge
libgcc                    7.2.0                h69d50b8_2    conda-forge
libgcc-ng                 8.2.0                hdf63c60_1  
libgfortran-ng            7.3.0                hdf63c60_0  
libiconv                  1.15              h516909a_1005    conda-forge
liblapack                 3.8.0                8_openblas    conda-forge
liblapacke                3.8.0                8_openblas    conda-forge
libopenblas               0.3.3                h5a2b251_3  
libpng                    1.6.37               hed695b0_0    conda-forge
libssh2                   1.8.2                h22169c7_2    conda-forge
libstdcxx-ng              8.2.0                hdf63c60_1  
libtiff                   4.0.10            h648cc4a_1001    conda-forge
libuuid                   2.32.1            h14c3975_1000    conda-forge
libxcb                    1.13              h14c3975_1002    conda-forge
libxml2                   2.9.9                h13577e0_0    conda-forge
libxslt                   1.1.32            h4785a14_1002    conda-forge
lxml                      4.3.3            py36h7ec2d77_0    conda-forge
matplotlib                2.2.3            py36hb69df0a_0  
matplotlib-base           2.2.3            py36h60b886d_1    conda-forge
mcl                       14.137          pl526h470a237_4    bioconda
megahit                   1.1.3                    py36_0    bioconda
mistune                   0.8.1            py36h3d5977c_0  
mkl                       2018.0.3                      1  
mmtf-python               1.1.2                      py_0    conda-forge
msgpack-python            0.6.1            py36h6bb024c_0    conda-forge
mummer                    3.23                    pl526_8    bioconda
muscle                    3.8.1551             h6bb024c_4    bioconda
ncurses                   6.1               hf484d3e_1002    conda-forge
nose                      1.3.7                 py36_1002    conda-forge
numpy                     1.16.3           py36he5ce36f_0    conda-forge
olefile                   0.46                       py_0    conda-forge
openblas                  0.3.6                h6e990d7_1    conda-forge
openssl                   1.1.1b               h14c3975_1    conda-forge
pandas                    0.24.2           py36hf484d3e_0    conda-forge
pango                     1.40.14           h4ea9474_1004    conda-forge
patsy                     0.5.1                      py_0    conda-forge
pcre                      8.41              hf484d3e_1003    conda-forge
perl                      5.26.2            h516909a_1006    conda-forge
pillow                    6.0.0            py36he7afcd5_0    conda-forge
pip                       19.1                     py36_0    conda-forge
pixman                    0.34.0            h14c3975_1003    conda-forge
prodigal                  2.6.3                         1    bioconda
psutil                    5.4.3                    py36_0    conda-forge
pthread-stubs             0.4               h14c3975_1001    conda-forge
pyani                     0.2.7            py36h24bf2e0_1    bioconda
pycparser                 2.19                     py36_1    conda-forge
pyopenssl                 19.0.0                   py36_0    conda-forge
pyparsing                 2.4.0                      py_0    conda-forge
pyqt                      5.9.2            py36h05f1152_2  
pyrsistent                0.15.1           py36h516909a_0    conda-forge
pysam                     0.15.2           py36hb06f55c_2    bioconda
pysocks                   1.6.8                 py36_1002    conda-forge
python                    3.6.7             h381d211_1004    conda-forge
python-dateutil           2.8.0                      py_0    conda-forge
python-levenshtein        0.12.0                   pypi_0    pypi
pytz                      2019.1                     py_0    conda-forge
pyyaml                    5.1              py36h14c3975_0    conda-forge
qt                        5.9.7                h52cfd70_1    conda-forge
ratelimiter               1.2.0                 py36_1000    conda-forge
readline                  7.0               hf8c457e_1001    conda-forge
reportlab                 3.5.20           py36h7d98c4e_0    conda-forge
requests                  2.20.0                py36_1000    conda-forge
samtools                  1.9                 h8571acd_11    bioconda
scikit-learn              0.19.2           py36h22eb022_0  
scipy                     1.2.1            py36h09a28d5_1    conda-forge
seaborn                   0.9.0                      py_1    conda-forge
setuptools                41.0.1                   py36_0    conda-forge
sip                       4.19.8          py36hf484d3e_1000    conda-forge
six                       1.11.0                py36_1001    conda-forge
snakemake-minimal         5.2.4                    py36_0    bioconda
sqlite                    3.26.0            h67949de_1001    conda-forge
statsmodels               0.9.0           py36h3010b51_1000    conda-forge
tabulate                  0.8.3                      py_0    conda-forge
tbb                       2019.5               hc9558a2_0    conda-forge
tk                        8.6.9             h84994c4_1001    conda-forge
tornado                   6.0.2            py36h516909a_0    conda-forge
trimal                    1.4.1                h6bb024c_3    bioconda
urllib3                   1.23                  py36_1001    conda-forge
wheel                     0.33.1                   py36_0    conda-forge
wrapt                     1.11.1           py36h14c3975_0    conda-forge
xorg-kbproto              1.0.7             h14c3975_1002    conda-forge
xorg-libice               1.0.9             h516909a_1004    conda-forge
xorg-libsm                1.2.3             h84519dc_1000    conda-forge
xorg-libx11               1.6.7             h14c3975_1000    conda-forge
xorg-libxau               1.0.9                h14c3975_0    conda-forge
xorg-libxdmcp             1.1.3                h516909a_0    conda-forge
xorg-libxext              1.3.4                h516909a_0    conda-forge
xorg-libxrender           0.9.10            h516909a_1002    conda-forge
xorg-libxt                1.1.5             h14c3975_1002    conda-forge
xorg-renderproto          0.11.1            h14c3975_1002    conda-forge
xorg-xextproto            7.3.0             h14c3975_1002    conda-forge
xorg-xproto               7.0.31            h14c3975_1007    conda-forge
xz                        5.2.4             h14c3975_1001    conda-forge
yaml                      0.1.7             h14c3975_1001    conda-forge
zlib                      1.2.11            h14c3975_1004    conda-forge
```

**Comment by ozcan** [comment]
Hi,
It seems for some reason `setup.py` can not find the gsl. you can manually put gsl paths in `setup.py`.

There is a section in setup.py like below:
```
include_dirs_for_concoct = [numpy.get_include(), '/opt/local/include/', '/usr/local/include/']
library_dirs_for_concoct = ['/opt/local/lib/', '/usr/local/lib/']
```
In order to find location of gsl on your system, you can use `pkg-config` tool.
```pkg-config --libs-only-L --cflags-only-I  gsl``` 
should print something like below (the one with -I is includes directory, -L is the library)
```
-I/usr/local/Cellar/gsl/2.5/include -L/usr/local/Cellar/gsl/2.5/lib
```

Best,

**Comment by michaelkyu** [comment]
I was running into the same build error with gsl. I installed `gsl` with conda, but `pkg-config` still could not find it. 

I was able to solve this by installing `gcc` with `conda install gcc`. I'm not sure why this works, but I got the idea from https://github.com/BinPro/CONCOCT/issues/117#issuecomment-343622462

**Comment by ozcan** [comment]
Awesome solution @michaelkyu,
It actually makes perfect sense to me. I did not realize until your comment but, the `pkg-config` output on my comment has paths with `/usr/local/Cellar/` prefix, which is a `homebrew` package manager directory. So I was using homebrew's gcc/pkg-config/gsl for compiling.

After your comment, I tried installing `conda install pkg-config` which correctly printed.
```
-I/Users/ozcan/miniconda3/include -L/Users/ozcan/miniconda3/lib
```

It seems conda's gcc already has these directories by default for searching libraries.

Best,
